### PR TITLE
Fix local index deletion timing

### PR DESCRIPTION
**Pull request by dzane17**
### Description
Fix local index deletion
- Run job every day at 00:05 UTC instead of when `QueryInsightsService` was launched
- Use 00:00 as expiration cutoff instead of `current_time - delete_after_days`
- Check & delete expired indices on cluster manager node only instead of all nodes

### Testing
Added logs to validate:
1. `getInitialDelay()` returns correct duration
2. `deleteExpiredTopNIndices()` is called at 00:05 UTC

```
[2025-04-08T23:59:24,998][INFO ][o.o.p.i.c.s.QueryInsightsService] [integTest-0] scheduling job
[2025-04-08T23:59:24,999][INFO ][o.o.p.i.c.s.QueryInsightsService] [integTest-0] in getInitialDelay
[2025-04-08T23:59:24,999][INFO ][o.o.p.i.c.s.QueryInsightsService] [integTest-0] delay = 35000
[2025-04-08T23:59:24,999][INFO ][o.o.p.i.c.s.QueryInsightsService] [integTest-0] scheduled job
...
[2025-04-09T00:05:00,000][INFO ][o.o.p.i.c.s.QueryInsightsService] [integTest-0] starting deleteExpiredTopNIndices
```

### Issues Resolved
Resolves https://github.com/opensearch-project/query-insights/issues/288

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by deshsidd** [comment]
LGTM

File: src/test/java/org/opensearch/plugin/insights/core/service/QueryInsightsServiceTests.java [hunk 1]
```diff
@@ -559,6 +562,20 @@ public void testSetExporterAndReaderType_CloseWithException() throws IOException
...
+        Instant instantNormal = Instant.parse("2025-03-26T10:15:30Z");
+        Instant startOfNextDayNormal = instantNormal.truncatedTo(ChronoUnit.DAYS).plus(1, ChronoUnit.DAYS);
+        final long expectedDelayNormal = Duration.between(instantNormal, startOfNextDayNormal).toMillis();
+        assertEquals(expectedDelayNormal, getInitialDelay(instantNormal).millis());
+
+        //// Second Test Case (Edge Case: Midnight UTC) ////
+        Instant instantEdge = Instant.parse("2025-03-26T00:00:00Z");
+        Instant startOfNextDayEdge = instantEdge.plus(1, ChronoUnit.DAYS);
+        final long expectedDelayEdge = Duration.between(instantEdge, startOfNextDayEdge).toMillis();
+        assertEquals(expectedDelayEdge, getInitialDelay(instantEdge).millis());
```
Comment (ansjcy) [inline src/test/java/org/opensearch/plugin/insights/core/service/QueryInsightsServiceTests.java:576]: Is it possible to add tests for schedule after the first day by mocking the scheduler?

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java [hunk 2]
```diff
@@ -567,6 +572,41 @@ public void validateTemplatePriority(final long templatePriority) {
...
     }
 
+    /**
+     * Schedules a daily task to delete expired Top N indices.
+     *
+     * @see #deleteExpiredTopNIndices()
+     */
+    private void createDailyDeleteAfterJob() {
+        // Run job once now
+        deleteExpiredTopNIndices();
```
Comment (ansjcy) [inline src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java:582]: Could you add some error handling here to avoid crashing the process if error happens? If `deleteExpiredTopNIndices` somehow error out, we should also be able to add the schedule job.

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java [hunk 3]
```diff
@@ -567,6 +572,41 @@ public void validateTemplatePriority(final long templatePriority) {
...
+
+        // Schedule recurring job with 1-day interval
+        scheduledFutures.add(
```
Comment (ansjcy) [inline src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java:585]: Just had a thought, this QueryInsightsService runs on every node, we're basically spinning up N deletion jobs right now. This is an issue in the currently implementation as well. We should really just create 1 job here.

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java [hunk 4]
```diff
@@ -567,6 +572,41 @@ public void validateTemplatePriority(final long templatePriority) {
...
+     */
+    static TimeValue getInitialDelay(final Instant now) {
+        // Calculate the start of the next UTC day (00:00:00)
+        final Instant startOfNextDay = now.truncatedTo(ChronoUnit.DAYS).plus(1, ChronoUnit.DAYS);
+        final long delayInMillis = Duration.between(now, startOfNextDay).toMillis();
+        return TimeValue.timeValueMillis(delayInMillis);
+    }
+
     @Override
     protected void doStart() {
```
Comment (ansjcy) [inline src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java:611]: Looks like we don't have any unit tests for the doStart method yet. Since we're adding this createDailyDeleteAfterJob stuff which makes things more complex, we should probably add some tests to cover it. That way we can make sure threadPool.schedule gets called properly, and if someone changes something later that breaks it, the tests will catch it.

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java [hunk 5]
```diff
@@ -578,12 +618,12 @@ protected void doStart() {
...
-                threadPool.scheduleWithFixedDelay(
-                    this::deleteExpiredTopNIndices,
-                    new TimeValue(1, TimeUnit.DAYS), // Check for deletable indices once per day
-                    QueryInsightsSettings.QUERY_INSIGHTS_EXECUTOR
-                )
+
+            // One-time job to run at next UTC 00:00
+            threadPool.schedule(
+                this::createDailyDeleteAfterJob,
+                getInitialDelay(Instant.now()),
```
Comment (ansjcy) [inline src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java:625]: The potential problem of this approach is, think about 2 cases (say today is April 2nd and delete after is set to 1 day):
- if the process is started at 23:59:59:98, it's possible `createDailyDeleteAfterJob` will be scheduled on 00:00 on the next day, then deleteExpiredTopNIndices will delete everything before April 1
- if the process is started at 23:59:59:98, it's also possible `createDailyDeleteAfterJob` will be scheduled on 23:59:59:99 on the same day, then deleteExpiredTopNIndices will delete everything before March 31.

Then the behavior might be inconsistent. Instead of `// One-time job to run at next UTC 00:00`, can we schedule it at some other time to make the behavior consistent?

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java [hunk 6]
```diff
@@ -567,6 +572,41 @@ public void validateTemplatePriority(final long templatePriority) {
...
+     *
+     * @see #deleteExpiredTopNIndices()
+     */
+    private void createDailyDeleteAfterJob() {
+        // Run job once now
+        deleteExpiredTopNIndices();
+
+        // Schedule recurring job with 1-day interval
+        scheduledFutures.add(
+            threadPool.scheduleWithFixedDelay(
```
Comment (ansjcy) [inline src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java:586]: We need tests to make sure this path is called by mocking the system time.

**Review by ansjcy — CHANGES_REQUESTED** [review]

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java (hunk 2 above)
Comment (dzane17) [reply src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java]: Added try/catch around `deleteExpiredTopNIndices()`

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java (hunk 4 above)
Comment (dzane17) [reply src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java:623]: Added a test for doStart

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java (hunk 5 above)
Comment (dzane17) [reply src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java]: Changed execution time from 00:00 to 00:05. That way we are clear of any rounding/millisecond errors that could cause it to run on the previous day.

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java (hunk 3 above)
Comment (dzane17) [reply src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java]: Good point. We should only schedule the delete job on the cluster manager node. With that said, we need to ensure the job stops/starts anytime the following events occur. 1-2 are already handled in the `QueryInsightsService` `doStart()` method. For number 3 I have added a listener which is called anytime the cluster manager node changes.
1. QI process restart
2. All top N & search metrics features toggle off/on
3. Cluster manager node changes

Another piece of complexity is that `QueryInsightsService` is created before `clusterService.clusterApplierService.state()` is initialized. This means the cluster manager node is unknown when QI process starts. Deletion job in this case will be scheduled by the listener once state is defined. If this order were to change, we are still okay. The job will be scheduled in `QueryInsightsService.doStart()` and the listener will pass since there is already an active ScheduledFuture present.

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java (hunk 6 above)
Comment (dzane17) [reply src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java]: New `testDoStart()` checks a ScheduledFuture is created.

File: src/test/java/org/opensearch/plugin/insights/core/service/QueryInsightsServiceTests.java (hunk 1 above)
Comment (dzane17) [reply src/test/java/org/opensearch/plugin/insights/core/service/QueryInsightsServiceTests.java]: I don't see a way to truly mock the passage of time and ensure `deleteExpiredTopNIndices` is called at a certain time by the threadpool scheduler.

We have tests for `getInitialDelay` and to ensure a ScheduledFuture object is created. The only missing piece is to validate `scheduleWithFixedDelay()` executes as expected. This is a java util so I think it is a must-have to write our own tests considering it may not be viable.

If anyone knows a way I am happy to add it.

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java (hunk 3 above)
Comment (dzane17) [reply src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java]: Actually the `LifecycleListener` I was using will not work. It is only called when the ClusterManagerService class is created/killed. It will not always trigger when the cluster manager node changes.

I have reverted back to scheduling a job on all nodes, but early return on all nodes except the cluster manager. This is a simpler approach since we don't need to worry about the cluster manager changing and it adds little overhead - 1 call to clusterService per node per day.

File: src/test/java/org/opensearch/plugin/insights/core/service/QueryInsightsServiceTests.java (hunk 1 above)
Comment (ansjcy) [reply src/test/java/org/opensearch/plugin/insights/core/service/QueryInsightsServiceTests.java]: Since we are using the `threadPool.scheduler().scheduleWithFixedDelay` function now, I'm good with the current tests.

**Review by ansjcy — COMMENTED** [review]
Looks good to me!

**Review by ansjcy — APPROVED** [review]

File: src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java (hunk 3 above)
Comment (ansjcy) [reply src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java]: this ^ makes sense to me.

**Comment by ansjcy** [comment]
Have we tested this logic
```
clusterService.localNode().getRoles().contains(DiscoveryNodeRole.CLUSTER_MANAGER_ROLE)) {
                            deleteExpiredTopNIndices();
                        }
```
On a multi node cluster? If not we should do it later.

**Comment by dzane17** [comment]
> On a multi node cluster? If not we should do it later.

Not yet. I will log this as a todo

**Comment by dzane17** [comment]
I tested on a 3 node domain. Turns out we need to use **elected** cluster manager instead of cluster manager role. I have updated to `isLocalNodeElectedClusterManager()`. 

These logs show all 3 nodes return `true` for `isClusterManagerNode()` but only 1 node is the elected cluster manager.
```
[2025-04-09T18:59:46,145][INFO ][o.o.p.i.c.s.QueryInsightsService] [integTest-0] isLocalNodeElectedClusterManager=false
[2025-04-09T18:59:46,147][INFO ][o.o.p.i.c.s.QueryInsightsService] [integTest-1] isLocalNodeElectedClusterManager=true
[2025-04-09T18:59:46,146][INFO ][o.o.p.i.c.s.QueryInsightsService] [integTest-2] isLocalNodeElectedClusterManager=false
[2025-04-09T18:59:46,146][INFO ][o.o.p.i.c.s.QueryInsightsService] [integTest-0] isClusterManagerNode=true
[2025-04-09T18:59:46,148][INFO ][o.o.p.i.c.s.QueryInsightsService] [integTest-1] isClusterManagerNode=true
[2025-04-09T18:59:46,147][INFO ][o.o.p.i.c.s.QueryInsightsService] [integTest-2] isClusterManagerNode=true
```

**Comment by ansjcy** [comment]
>  Turns out we need to use elected cluster manager instead of cluster manager role.

Interesting, I assume there would only be one elected cluster manager on a running cluster (one node/multiple nodes scenarios)?

**Comment by dzane17** [comment]
Yes, there is only one elected cluster manager at a time. The `CLUSTER_MANAGER_ROLE` is assigned to all nodes that are eligible to be cluster manager.

Ref: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/node/DiscoveryNodeRole.java#L244-L247
